File: Algorithms/Easy/MoveZeroes.py
# Time complexity: O(n)
class Solution:
    # Space complexity: O(n)
    def moveZeroes(self, nums: list[int]) -> None:
        nums_cp = nums[:]

        i, j = 0, len(nums) - 1
        for num in nums_cp:
            if num != 0:
                nums[i] = num
                i += 1
            else:
                nums[j] = num
                j -= 1

    # Space complexity: O(1)
    def moveZeroes_1(self, nums: list[int]) -> None:
        slow = 0
        for fast in range(len(nums)):
            if nums[fast]:
                nums[slow], nums[fast] = nums[fast], nums[slow]
                slow += 1

File: Algorithms/Easy/283.Move-Zeroes/MoveZeroes.go
package easy

// Time complexity: O(n)
// Space complexity: O(1)
func moveZeroes_1(nums []int) {
	lastNonZero := 0
	for i := range nums {
		if nums[i] != 0 {
			nums[i], nums[lastNonZero] = nums[lastNonZero], nums[i]
			lastNonZero++
		}
	}
}

// Time complexity: O(n)
// Space complexity: O(n)
func moveZeroes_2(nums []int) {
	numsCp := make([]int, len(nums))
	copy(numsCp, nums)
	l, r := 0, len(nums)-1
	for _, num := range numsCp {
		if num != 0 {
			nums[l] = num
			l++
		} else {
			nums[r] = num
			r--
		}
	}
}
